Fix overflows in the implementation of `overflowing_literals` lint's help

s-cerevisiae opened the pull request:
This PR fixes two overflow problems that cause the `overflowing_literals` lint to behave incorrectly in some edge cases.

1. When an integer literal is between `i128::MAX` and `u128::MAX`, an overflowing `as` cast can cause the suggested type to be overly small. It's fixed by using checked type conversion and returning `u128` when it's the only choice. (Fixes #135248)
2. When an integer literal is `i128::MIN` but is of a smaller type, an overflowing negation cause the compiler to panic in debug build. Fixed by checking the number size beforehand and `wrapping_neg`. (Fixes #131849)

Edit: extracted the type conversion part into a standalone function to separate the concern of overflowing.

s-cerevisiae committed:
Detect overflow when the literal is larger than i128::MAX

s-cerevisiae committed:
Detect overflow when the literal is negative

s-cerevisiae committed:
Add test cases and test for `HELP`

s-cerevisiae committed:
Extract integer conversion into a function

s-cerevisiae committed:
Minor simplification

Apply eta-reduction on map to simplify code and make the style more
consistent

chenyukang commented on tests/ui/lint/type-overflow.rs:
```diff
@@ -3,20 +3,46 @@
+    //~| HELP consider using the type `u128` instead
+    //~| HELP consider using the type `u128` for the literal and cast it to `i128`
+
+    let fail = 0x8000_0000_0000_0000_0000_0000_0000_0000;
```
it's better to add comments to mark this case is from: 
#131849

s-cerevisiae replied:
Done, thanks for the suggestion.

chenyukang commented:
Please add:
```
Fixes #135399 
Fixes #131849
```
in the PR description so that the two issue will be closed after PR merged.

chenyukang commented:
seems we does not find the proper type for different range of number literals , for example:

```
let _ = 18446744073709551614;
```
the value is `u64::MAX - 1`, the current suggestion is:

```console
  |
3 |     let _ = 18446744073709551614;
  |             ^^^^^^^^^^^^^^^^^^^^
  |
  = note: the literal `18446744073709551614` does not fit into the type `i32` whose range is `-2147483648..=2147483647`
  = help: consider using the type `i128` instead
```

maybe better choice is suggesting `u64`.

s-cerevisiae committed:
Add comment on case to mark the original issue

s-cerevisiae commented:
@chenyukang I preserved the original behavior which is to suggest types with the same signedness unless the value fits into an unsigned number of the same size. I was rather reluctant to change the suggestion logic in this PR, but if you feel that it's ok to do here I'll change it so that it always suggests the smallest fitting type.

s-cerevisiae committed:
Suggest the smallest fitting type instead

Changes the behavior of the `overflowing_literals` suggestion so that it
always suggest the smallest type regardless of the original type size.

s-cerevisiae commented:
Done implementing that too. Hope that there was no special reason for the old, less straightforward behavior.

chenyukang commented:
@bors r+ rollup

bors commented:
:pushpin: Commit 74e2e8b59834aede7aafee57e00ede07ae13739b has been approved by `chenyukang`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=chenyukang 74e2e8b59834aede7aafee57e00ede07ae13739b -->
<!-- homu: {"type":"Approved","sha":"74e2e8b59834aede7aafee57e00ede07ae13739b","approver":"chenyukang","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
